Server tests fail intermittently due to spark.stop() destroying shared SparkContext

## Description

Several server tests fail intermittently with Spark session errors when the full test suite is run. The failures are non-deterministic — different tests fail on different runs — because they depend on which tests happen to execute first.

## Failing tests

- `ViewDefinitionSearchTest` — 9 errors
- `ViewDefinitionCreateTest` — 6 errors

## Error extract

```
[ERROR] Tests run: 9, Failures: 0, Errors: 9, Skipped: 0, Time elapsed: 2.911 s <<< FAILURE! -- in au.csiro.pathling.operations.view.ViewDefinitionSearchTest
[ERROR] au.csiro.pathling.operations.view.ViewDefinitionSearchTest.searchViewDefinitionsWithNameFilter -- Time elapsed: 0.320 s <<< ERROR!
org.apache.spark.SparkException: [INTERNAL_ERROR] The SparkPlan.session() method is called but the associated SparkSession is null.

[ERROR] Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 0.296 s <<< FAILURE! -- in au.csiro.pathling.operations.view.ViewDefinitionCreateTest
[ERROR] au.csiro.pathling.operations.view.ViewDefinitionCreateTest.createViewDefinitionGeneratesNewUuid -- Time elapsed: 0.296 s <<< ERROR!
org.apache.spark.SparkException: [INTERNAL_ERROR] No active or default Spark session found SQLSTATE: XX000
	at org.apache.spark.SparkException$.internalError(SparkException.scala:92)
	at org.apache.spark.SparkException$.internalError(SparkException.scala:96)
	at org.apache.spark.sql.SparkSessionCompanion.$anonfun$active$2(SparkSession.scala:1031)
	at scala.Option.getOrElse(Option.scala:201)
	at org.apache.spark.sql.SparkSessionCompanion.$anonfun$active$1(SparkSession.scala:1031)
	at scala.Option.getOrElse(Option.scala:201)
	at org.apache.spark.sql.SparkSessionCompanion.active(SparkSession.scala:1030)
	at org.apache.spark.sql.delta.DeltaTableUtils$.findDeltaTableRoot(DeltaTable.scala:206)
	at org.apache.spark.sql.delta.DeltaTableUtils$.findDeltaTableRoot(DeltaTable.scala:198)
	at io.delta.tables.DeltaTable$.isDeltaTable(DeltaTable.scala:1283)
	at io.delta.tables.DeltaTable.isDeltaTable(DeltaTable.scala)
	at au.csiro.pathling.operations.update.UpdateExecutor.deltaTableExists(UpdateExecutor.java:243)
	at au.csiro.pathling.operations.update.UpdateExecutor.merge(UpdateExecutor.java:145)
	at au.csiro.pathling.operations.update.UpdateExecutor.merge(UpdateExecutor.java:121)
	at au.csiro.pathling.operations.create.CreateProvider.create(CreateProvider.java:111)
	at au.csiro.pathling.operations.view.ViewDefinitionCreateTest.createViewDefinitionGeneratesNewUuid(ViewDefinitionCreateTest.java:119)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

## Root cause

Three standalone tests created their own `SparkSession` via `SparkSession.builder().getOrCreate()` and called `spark.stop()` in an `@AfterAll` lifecycle method:

- `SqlQueryResultStreamerTest`
- `ViewRegistrationServiceTest`
- `LibraryReferenceResolverTest.CanonicalReferences`

Because `getOrCreate()` returns the JVM-wide singleton `SparkSession`, calling `spark.stop()` on any instance destroys the shared `SparkContext` for the entire JVM process. Any test that runs afterward and tries to use Spark fails with "no active or default Spark session found" or a null `SparkPlan.session()`.

The failures were non-deterministic because they only occurred when one of the three offending tests happened to run before `ViewDefinitionSearchTest` or `ViewDefinitionCreateTest` within the same JVM.

## Fix

Removed the `@AfterAll spark.stop()` calls from the three standalone tests and converted them to proper `@SpringBootUnitTest` tests that receive the shared `SparkSession` via Spring injection, consistent with all other Spark-dependent tests in the server module. The Spring-managed session is never stopped by individual tests.

--- a/server/src/main/java/au/csiro/pathling/operations/bulkimport/ImportPnpExecutor.java
+++ b/server/src/main/java/au/csiro/pathling/operations/bulkimport/ImportPnpExecutor.java
@@ -158,9 +158,12 @@ public class ImportPnpExecutor {
 
       // Execute the import using the existing ImportExecutor with custom allowable sources.
       // This bypasses the configured allowableSources validation for the staging directory,
-      // which the server downloaded and trusts. The qualified URI ensures the prefix matches
-      // whatever scheme the staging file system uses.
-      final List<String> pnpAllowableSources = List.of(tempDir.toUri().toString() + "/");
+      // which the server downloaded and trusts. Go via fs.getFileStatus() to obtain a URI in
+      // the same canonical form (with empty authority preserved on file://) that fs.listFiles
+      // produces for the downloaded files, so the UrlAllowlist string-prefix match holds:
+      // tempDir.toUri() alone yields file:/path while listed files come back as file:///path.
+      final List<String> pnpAllowableSources =
+          List.of(fs.getFileStatus(tempDir).getPath().toUri().toString() + "/");
       final ImportResponse response =
           importExecutor.execute(importRequest, jobId, pnpAllowableSources);
 

--- a/server/src/test/java/au/csiro/pathling/operations/bulkimport/ImportPnpOperationIT.java
+++ b/server/src/test/java/au/csiro/pathling/operations/bulkimport/ImportPnpOperationIT.java
@@ -701,6 +701,7 @@ class ImportPnpOperationIT {
             .header("Content-Type", "application/fhir+json")
             .header("Accept", "application/fhir+json")
             .header("Prefer", "respond-async")
+            .header("Authorization", "Bearer " + AUTH_TOKEN)
             .bodyValue(requestBody)
             .exchange()
             .expectStatus()
@@ -736,6 +737,7 @@ class ImportPnpOperationIT {
                     .get()
                     .uri(contentLocation)
                     .header("Accept", "application/fhir+json")
+                    .header("Authorization", "Bearer " + AUTH_TOKEN)
                     .exchange()
                     .expectStatus()
                     .is4xxClientError()
@@ -752,6 +754,7 @@ class ImportPnpOperationIT {
     webTestClient
         .get()
         .uri("http://localhost:" + port + "/jobs/" + jobId + "/escaped.0000.ndjson")
+        .header("Authorization", "Bearer " + AUTH_TOKEN)
         .exchange()
         .expectStatus()
         .isNotFound();

--- a/server/src/test/java/au/csiro/pathling/operations/bulkimport/ImportPnpOperationIT.java
+++ b/server/src/test/java/au/csiro/pathling/operations/bulkimport/ImportPnpOperationIT.java
@@ -173,6 +173,11 @@ class ImportPnpOperationIT {
 
   @AfterEach
   void cleanup() throws IOException {
+    // Clear cached Delta table state before deleting files. Otherwise the next test sees a stale
+    // DeltaLog in memory that no longer matches the on-disk warehouse rebuilt from test fixtures,
+    // and Delta refuses the import with DELTA_PATH_EXISTS.
+    pathlingContext.getSpark().catalog().clearCache();
+    org.apache.spark.sql.delta.DeltaLog.clearCache();
     FileUtils.cleanDirectory(warehouseDir.toFile());
   }
 

--- a/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitOAuthIT.java
+++ b/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitOAuthIT.java
@@ -71,7 +71,6 @@ import org.springframework.test.web.reactive.server.WebTestClient;
 @TestPropertySource(
     properties = {
       "pathling.async.enabled=true",
-      "pathling.bulk-submit.allowable-sources[0]=http://localhost",
       // Configure submitter with OAuth credentials for symmetric (client_secret) auth.
       "pathling.bulk-submit.allowed-submitters[0].system=http://example.org/submitters",
       "pathling.bulk-submit.allowed-submitters[0].value=oauth-submitter",
@@ -118,6 +117,11 @@ class BulkSubmitOAuthIT {
     TestDataSetup.copyTestDataToTempDir(warehouseDir, "Condition");
     registry.add("pathling.storage.warehouseUrl", () -> "file://" + warehouseDir.toAbsolutePath());
     registry.add("pathling.bulk-submit.staging-directory", stagingDir::toString);
+    // The allowable source must include the WireMock port. Bare "http://localhost" (port 80)
+    // no longer matches "http://localhost:{port}" under the URI-aware UrlAllowlist matching.
+    registry.add(
+        "pathling.bulk-submit.allowable-sources[0]",
+        () -> "http://localhost:" + wireMockServer.port());
   }
 
   @BeforeEach

--- a/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitOperationIT.java
+++ b/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitOperationIT.java
@@ -67,7 +67,6 @@ import org.springframework.test.web.reactive.server.WebTestClient;
 @TestPropertySource(
     properties = {
       "pathling.async.enabled=true",
-      "pathling.bulk-submit.allowable-sources[0]=http://localhost",
       "pathling.bulk-submit.allowed-submitters[0].system=http://example.org/submitters",
       "pathling.bulk-submit.allowed-submitters[0].value=test-submitter"
     })
@@ -107,6 +106,11 @@ class BulkSubmitOperationIT {
     TestDataSetup.copyTestDataToTempDir(warehouseDir, "Condition");
     registry.add("pathling.storage.warehouseUrl", () -> "file://" + warehouseDir.toAbsolutePath());
     registry.add("pathling.bulk-submit.staging-directory", stagingDir::toString);
+    // The allowable source must include the WireMock port. Bare "http://localhost" (port 80)
+    // no longer matches "http://localhost:{port}" under the URI-aware UrlAllowlist matching.
+    registry.add(
+        "pathling.bulk-submit.allowable-sources[0]",
+        () -> "http://localhost:" + wireMockServer.port());
   }
 
   @BeforeEach

--- a/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitProviderTest.java
+++ b/server/src/test/java/au/csiro/pathling/operations/bulksubmit/BulkSubmitProviderTest.java
@@ -34,6 +34,7 @@ import java.util.List;
 import java.util.Optional;
 import org.hl7.fhir.r4.model.Parameters;
 import org.hl7.fhir.r4.model.StringType;
+import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.springframework.security.core.Authentication;
@@ -79,6 +80,11 @@ class BulkSubmitProviderTest {
     final Authentication authentication = mock(Authentication.class);
     when(securityContext.getAuthentication()).thenReturn(authentication);
     SecurityContextHolder.setContext(securityContext);
+  }
+
+  @AfterEach
+  void tearDown() {
+    SecurityContextHolder.clearContext();
   }
 
   // ========================================

--- a/server/src/test/java/au/csiro/pathling/operations/sqlquery/LibraryReferenceResolverTest.java
+++ b/server/src/test/java/au/csiro/pathling/operations/sqlquery/LibraryReferenceResolverTest.java
@@ -28,6 +28,7 @@ import au.csiro.pathling.encoders.FhirEncoders;
 import au.csiro.pathling.errors.ResourceNotFoundError;
 import au.csiro.pathling.io.source.DataSource;
 import au.csiro.pathling.read.ReadExecutor;
+import au.csiro.pathling.test.SpringBootUnitTest;
 import ca.uhn.fhir.rest.server.exceptions.InvalidRequestException;
 import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
 import java.util.List;
@@ -38,21 +39,18 @@ import org.hl7.fhir.instance.model.api.IBaseResource;
 import org.hl7.fhir.r4.model.Enumerations.PublicationStatus;
 import org.hl7.fhir.r4.model.Library;
 import org.hl7.fhir.r4.model.Reference;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.TestInstance;
-import org.junit.jupiter.api.TestInstance.Lifecycle;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.CsvSource;
+import org.springframework.beans.factory.annotation.Autowired;
 
 /**
  * Tests for {@link LibraryReferenceResolver} covering both the relative-literal and canonical
  * reference resolution paths.
  */
-@TestInstance(Lifecycle.PER_CLASS)
+@SpringBootUnitTest
 class LibraryReferenceResolverTest {
 
   // ---------------------------------------------------------------------------
@@ -127,38 +125,17 @@ class LibraryReferenceResolverTest {
   }
 
   // ---------------------------------------------------------------------------
-  // Canonical references — uses a real Spark session + FhirEncoders.
+  // Canonical references — uses the shared Spark session and FhirEncoders.
   // ---------------------------------------------------------------------------
 
   @Nested
-  @TestInstance(Lifecycle.PER_CLASS)
   class CanonicalReferences {
 
-    private SparkSession spark;
-    private FhirEncoders fhirEncoders;
+    @Autowired private SparkSession spark;
+    @Autowired private FhirEncoders fhirEncoders;
+
     private DataSource dataSource;
     private LibraryReferenceResolver resolver;
-
-    @BeforeAll
-    void setUpAll() {
-      spark =
-          SparkSession.builder()
-              .master("local[1]")
-              .appName("LibraryReferenceResolverTest")
-              .config("spark.driver.bindAddress", "localhost")
-              .config("spark.driver.host", "localhost")
-              .config("spark.ui.enabled", false)
-              .config("spark.sql.shuffle.partitions", 1)
-              .getOrCreate();
-      fhirEncoders = FhirEncoders.forR4().getOrCreate();
-    }
-
-    @AfterAll
-    void tearDownAll() {
-      if (spark != null) {
-        spark.stop();
-      }
-    }
 
     @BeforeEach
     void setUp() {

--- a/server/src/test/java/au/csiro/pathling/operations/sqlquery/SqlQueryResultStreamerTest.java
+++ b/server/src/test/java/au/csiro/pathling/operations/sqlquery/SqlQueryResultStreamerTest.java
@@ -19,6 +19,7 @@ package au.csiro.pathling.operations.sqlquery;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import au.csiro.pathling.test.SpringBootUnitTest;
 import java.nio.charset.StandardCharsets;
 import java.util.List;
 import org.apache.spark.sql.Dataset;
@@ -27,44 +28,21 @@ import org.apache.spark.sql.RowFactory;
 import org.apache.spark.sql.SparkSession;
 import org.apache.spark.sql.types.DataTypes;
 import org.apache.spark.sql.types.StructType;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.TestInstance;
-import org.junit.jupiter.api.TestInstance.Lifecycle;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.mock.web.MockHttpServletResponse;
 
 /**
- * Tests for {@link SqlQueryResultStreamer} covering each output format. Uses a real local
- * SparkSession to materialise a small Dataset and a Spring {@link MockHttpServletResponse} to
- * capture written bytes and headers.
+ * Tests for {@link SqlQueryResultStreamer} covering each output format. Uses the shared Spark
+ * session to materialise a small Dataset and a Spring {@link MockHttpServletResponse} to capture
+ * written bytes and headers.
  */
-@TestInstance(Lifecycle.PER_CLASS)
+@SpringBootUnitTest
 class SqlQueryResultStreamerTest {
 
-  private SparkSession spark;
-  private SqlQueryResultStreamer streamer;
+  @Autowired private SparkSession spark;
 
-  @BeforeAll
-  void setUpAll() {
-    spark =
-        SparkSession.builder()
-            .master("local[1]")
-            .appName("SqlQueryResultStreamerTest")
-            .config("spark.driver.bindAddress", "localhost")
-            .config("spark.driver.host", "localhost")
-            .config("spark.ui.enabled", false)
-            .config("spark.sql.shuffle.partitions", 1)
-            .getOrCreate();
-    streamer = new SqlQueryResultStreamer();
-  }
-
-  @AfterAll
-  void tearDownAll() {
-    if (spark != null) {
-      spark.stop();
-    }
-  }
+  private final SqlQueryResultStreamer streamer = new SqlQueryResultStreamer();
 
   @Test
   void streamsNdjsonWithUtf8Encoding() {

--- a/server/src/test/java/au/csiro/pathling/operations/sqlquery/SqlQueryRunDeltaIT.java
+++ b/server/src/test/java/au/csiro/pathling/operations/sqlquery/SqlQueryRunDeltaIT.java
@@ -29,6 +29,7 @@ import jakarta.annotation.Nonnull;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -106,6 +107,7 @@ class SqlQueryRunDeltaIT {
         webTestClient
             .mutate()
             .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(100 * 1024 * 1024))
+            .responseTimeout(Duration.ofSeconds(60))
             .build();
     jsonParser = fhirContext.newJsonParser();
   }

--- a/server/src/test/java/au/csiro/pathling/operations/sqlquery/ViewRegistrationServiceTest.java
+++ b/server/src/test/java/au/csiro/pathling/operations/sqlquery/ViewRegistrationServiceTest.java
@@ -20,6 +20,7 @@ package au.csiro.pathling.operations.sqlquery;
 import static org.assertj.core.api.Assertions.assertThat;
 
 import au.csiro.pathling.config.ServerConfiguration;
+import au.csiro.pathling.test.SpringBootUnitTest;
 import ca.uhn.fhir.context.FhirContext;
 import java.util.Arrays;
 import java.util.List;
@@ -35,42 +36,26 @@ import org.apache.spark.sql.RowFactory;
 import org.apache.spark.sql.SparkSession;
 import org.apache.spark.sql.types.DataTypes;
 import org.apache.spark.sql.types.StructType;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.TestInstance;
-import org.junit.jupiter.api.TestInstance.Lifecycle;
+import org.springframework.beans.factory.annotation.Autowired;
 
 /**
  * Tests for {@link ViewRegistrationService}, with particular attention to the request-id
  * namespacing that prevents concurrent {@code $sqlquery-run} requests from clobbering one another's
  * temporary views in Spark's session-global catalog.
  */
-@TestInstance(Lifecycle.PER_CLASS)
+@SpringBootUnitTest
 class ViewRegistrationServiceTest {
 
-  private SparkSession spark;
+  @Autowired private SparkSession spark;
+  @Autowired private FhirContext fhirContext;
+
   private ViewRegistrationService service;
 
-  @BeforeAll
+  @BeforeEach
   void setUp() {
-    spark =
-        SparkSession.builder()
-            .master("local[2]")
-            .appName("ViewRegistrationServiceTest")
-            .config("spark.driver.bindAddress", "localhost")
-            .config("spark.driver.host", "localhost")
-            .config("spark.ui.enabled", false)
-            .config("spark.sql.shuffle.partitions", 1)
-            .getOrCreate();
-    service = new ViewRegistrationService(spark, FhirContext.forR4(), new ServerConfiguration());
-  }
-
-  @AfterAll
-  void tearDown() {
-    if (spark != null) {
-      spark.stop();
-    }
+    service = new ViewRegistrationService(spark, fhirContext, new ServerConfiguration());
   }
 
   // ---------------------------------------------------------------------------

